rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,122 @@
+   * `%d` casts the number into `long long`, which has a lower max value than `double` and does not support decimals.
+   * `%f` by default will format to the millionths, e.g. `5.5` is `5.500000`.
+   * `%g` by default will format up to the hundred thousandths, e.g. `5.5` is `5.5` and `5.5312389` is `5.53123`. It will also convert the number to scientific notation when it encounters a number equal to or greater than 10^6.
+   * To not lose precision, you need to use `%s`, but even so the type checker assumes you actually wanted strings.
+3. No support for `boolean`. You must use `%s` **and** call `tostring`.
+4. No support for values implementing the `__tostring` metamethod.
+5. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix issue #5, we will need new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting style may follow the expression, delimited by `,`.
+   * Restriction: the formatting style must be constant at parse time.
+   * The formatting style will follow the same syntax as pre-existing syntax in `string.format`.
+
+To put the above into formal EBNF grammar:
+
+```
+fmtflags ::= `-' | `+' | ` ' | `#' | `0'
+fmtwidth ::= <NUMBER> [<NUMBER>]
+fmtprecision ::= `.' <NUMBER> [<NUMBER>]
+fmtspecifier ::= `c' | `d' | `i' | `e' | `E' | `f' | `g' | `G' | `o' | `q' | `s' | `u' | `x' | `X'
+
+interpfmt ::= [fmtflags][fmtwidth][fmtprecision][fmtspecifier]
+
+interpexp ::= exp [`,' interpfmt]
+
+stringinterp ::= <INTERP_BEGIN> interpfmt {<INTERP_MID> interpexp} <INTERP_END>
+```
+
+Which, in actual Luau code, will look like the following:
+
+```
+local world = "world"
+print(`Hello {world}!`)
+--> Hello world!
+
+print(`0x{3735928559,X}`)
+--> 0xDEADBEEF
+
+local combo = {5, 2, 8, 9}
+print(`The lock combinations are: {table.concat(combo, ", ")}`)
+--> The lock combinations are: 5, 2, 8, 9
+```
+
+To fix issues #1, #2, #3, and #4, we will also want to extend `string.format` to accept braces after `%`. This is currently an invalid token, so this is a backward compatible extension. Unfortunately, formatting style cannot immediately follow `%` because `%d{}` is already valid and parses as `(%d){}`.
+
+The extended syntax will allow the following ``%{[<NUMBER>][`,' interpfmt]}`` to be parsed as one token.
+
+To put that into perspective:
+
+```lua
+print(string.format("%{} %{}", 1, 2))
+--> 1 2
+
+print(string.format("%{2} %{1}", 1, 2))
```

Comment:
`%{` would default to the `s` specifier, but also call `tostring`. Internally, it would just branch into the new formatting logic. The alternative is `%*` which is also unused and makes sense, but doesn't leave as much room for extensions.

Reply:
F# uses %A for this (I think A stands for Any). %a could also make sense since our any type is lowercase. %* could also work, although * is part of the options grammar in C we don't support it I think.

Reply:
`%a` and `%A` are used in C and newer versions of Lua for hexadecimal floats.
`%?` would work better than `%*` since it doesn't conflict with variable width which C supports.

Reply:
We would have to special case this logic anyway, so `%a` or `%*` is fine by me. I like `%a` more though.